File: Paintroid/src/main/java/org/catrobat/paintroid/contract/MainActivityContracts.java
/*
 * Paintroid: An image manipulation application for Android.
 * Copyright (C) 2010-2015 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.paintroid.contract;

import android.content.ContentResolver;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.util.DisplayMetrics;
import android.view.Menu;

import org.catrobat.paintroid.common.MainActivityConstants.ActivityRequestCode;
import org.catrobat.paintroid.dialog.PermissionInfoDialog;
import org.catrobat.paintroid.iotasks.CreateFileAsync;
import org.catrobat.paintroid.iotasks.LoadImageAsync;
import org.catrobat.paintroid.iotasks.SaveImageAsync;
import org.catrobat.paintroid.tools.ToolType;
import org.catrobat.paintroid.ui.LayerAdapter;

import java.io.File;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

public interface MainActivityContracts {
	interface Navigator {
		void showColorPickerDialog();

		void startLoadImageActivity(@ActivityRequestCode int requestCode);

		void startImportImageActivity(@ActivityRequestCode int requestCode);

		void showAboutDialog();

		void showLikeUsDialog();

		void showRateUsDialog();

		void showFeedbackDialog();

		void showOverwriteDialog(int permissionCode);

		void showPngInformationDialog();

		void showJpgInformationDialog();

		void sendFeedback();

		void startWelcomeActivity(@ActivityRequestCode int requestCode);

		void startShareImageActivity(Bitmap bitmap);

		void showIndeterminateProgressDialog();

		void dismissIndeterminateProgressDialog();

		void returnToPocketCode(String path);

		void showToast(@StringRes int resId, int duration);

		void showSaveErrorDialog();

		void showLoadErrorDialog();

		void showRequestPermissionRationaleDialog(PermissionInfoDialog.PermissionType permissionType, String[] permissions, int requestCode);

		void showRequestPermanentlyDeniedPermissionRationaleDialog();

		void askForPermission(String[] permissions, int requestCode);

		boolean isSdkAboveOrEqualM();

		boolean isSdkAboveOrEqualQ();

		boolean doIHavePermission(String permission);

		boolean isPermissionPermanentlyDenied(String[] permission);

		void finishActivity();

		void showSaveBeforeFinishDialog();

		void showSaveBeforeNewImageDialog();

		void showSaveBeforeLoadImageDialog();

		void showSaveImageInformationDialogWhenStandalone(int permissionCode, int imageNumber, boolean isExport);

		void restoreFragmentListeners();

		void showToolChangeToast(int offset, int idRes);

		void broadcastAddPictureToGallery(Uri uri);

		void rateUsClicked();

		void showImageImportDialog();

		void showCatroidMediaGallery();
	}

	interface MainView {
		Presenter getPresenter();

		boolean isFinishing();

		ContentResolver getContentResolver();

		DisplayMetrics getDisplayMetrics();

		void initializeActionBar(boolean isOpenedFromCatroid);

		void superHandleActivityResult(int requestCode, int resultCode, Intent data);

		void superHandleRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults);

		Uri getUriFromFile(File file);

		void hideKeyboard();

		boolean isKeyboardShown();

		void refreshDrawingSurface();

		void enterFullscreen();

		void exitFullscreen();
	}

	interface Presenter {
		void initializeFromCleanState(String extraPicturePath, String extraPictureName);

		void restoreState(boolean isFullscreen, boolean isSaved, boolean isOpenedFromCatroid,
				boolean wasInitialAnimationPlayed, @Nullable Uri savedPictureUri, @Nullable Uri cameraImageUri);

		void finishInitialize();

		void removeMoreOptionsItems(Menu menu);

		void loadImageClicked();

		void loadNewImage();

		void newImageClicked();

		void discardImageClicked();

		void saveCopyClicked(boolean isExport);

		void saveImageClicked();

		void shareImageClicked();

		void enterFullscreenClicked();

		void exitFullscreenClicked();

		void backToPocketCodeClicked();

		void showHelpClicked();

		void showAboutClicked();

		void showRateUsDialog();

		void showFeedbackDialog();

		void showOverwriteDialog(int permissionCode);

		void showPngInformationDialog();

		void showJpgInformationDialog();

		void sendFeedback();

		void onNewImage();

		void switchBetweenVersions(int requestCode);

		void handleActivityResult(int requestCode, int resultCode, Intent data);

		void handleRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults);

		void onBackPressed();

		void saveImageConfirmClicked(int requestCode, Uri uri);

		void undoClicked();

		void redoClicked();

		void showColorPickerClicked();

		void showLayerMenuClicked();

		void onCommandPostExecute();

		void setBottomNavigationColor(int color);

		void onCreateTool();

		void toolClicked(ToolType toolType);

		void saveBeforeLoadImage();

		void saveBeforeNewImage();

		void saveBeforeFinish();

		void finishActivity();

		void actionToolsClicked();

		void actionCurrentToolClicked();

		void rateUsClicked();

		void importFromGalleryClicked();

		void showImportDialog();

		void importStickersClicked();

		void bitmapLoadedFromSource(Bitmap loadedImage);

		void setLayerAdapter(LayerAdapter layerAdapter);

		int getImageNumber();

		Bitmap getBitmap();
	}

	interface Model {
		Uri getCameraImageUri();

		void setCameraImageUri(Uri cameraImageUri);

		Uri getSavedPictureUri();

		void setSavedPictureUri(Uri savedPictureUri);

		boolean isSaved();

		void setSaved(boolean saved);

		boolean isFullscreen();

		void setFullscreen(boolean fullscreen);

		boolean isOpenedFromCatroid();

		void setOpenedFromCatroid(boolean openedFromCatroid);

		boolean wasInitialAnimationPlayed();

		void setInitialAnimationPlayed(boolean wasInitialAnimationPlayed);
	}

	interface Interactor {
		void saveCopy(SaveImageAsync.SaveImageCallback callback, int requestCode, Bitmap bitmap);

		void createFile(CreateFileAsync.CreateFileCallback callback, int requestCode, @Nullable String filename);

		void saveImage(SaveImageAsync.SaveImageCallback callback, int requestCode, Bitmap bitmap, Uri uri);

		void loadFile(LoadImageAsync.LoadImageCallback callback, int requestCode, Uri uri);

		void loadFile(LoadImageAsync.LoadImageCallback callback, int requestCode, int maxWidth, int maxHeight, Uri uri);
	}

	interface TopBarViewHolder {
		void enableUndoButton();

		void disableUndoButton();

		void enableRedoButton();

		void disableRedoButton();

		void hide();

		void show();

		int getHeight();

		void removeStandaloneMenuItems(Menu menu);

		void removeCatroidMenuItems(Menu menu);

		void hideTitleIfNotStandalone();
	}

	interface DrawerLayoutViewHolder {

		void closeDrawer(int gravity, boolean animate);

		boolean isDrawerOpen(int gravity);

		void openDrawer(int gravity);
	}

	interface BottomBarViewHolder {
		void show();

		void hide();

		boolean isVisible();
	}

	interface BottomNavigationViewHolder {
		void show();

		void hide();

		void showCurrentTool(ToolType toolType);

		void setColorButtonColor(@ColorInt int color);
	}

	interface BottomNavigationAppearance {
		void showCurrentTool(ToolType toolType);
	}
}

File: Paintroid/src/main/java/org/catrobat/paintroid/contract/MainActivityContracts.kt
/*
 * Paintroid: An image manipulation application for Android.
 * Copyright (C) 2010-2022 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.catrobat.paintroid.contract

import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.util.DisplayMetrics
import android.view.Menu
import androidx.annotation.ColorInt
import androidx.annotation.StringRes
import org.catrobat.paintroid.UserPreferences
import org.catrobat.paintroid.command.serialization.CommandSerializer
import org.catrobat.paintroid.colorpicker.ColorHistory
import org.catrobat.paintroid.common.MainActivityConstants.ActivityRequestCode
import org.catrobat.paintroid.dialog.PermissionInfoDialog.PermissionType
import org.catrobat.paintroid.iotasks.CreateFile.CreateFileCallback
import org.catrobat.paintroid.iotasks.LoadImage.LoadImageCallback
import org.catrobat.paintroid.iotasks.SaveImage.SaveImageCallback
import org.catrobat.paintroid.iotasks.WorkspaceReturnValue
import org.catrobat.paintroid.tools.ToolType
import org.catrobat.paintroid.ui.LayerAdapter
import java.io.File

interface MainActivityContracts {
    interface Navigator {
        val isSdkAboveOrEqualM: Boolean
        val isSdkAboveOrEqualQ: Boolean

        fun showColorPickerDialog()

        fun startLoadImageActivity(@ActivityRequestCode requestCode: Int)

        fun startImportImageActivity(@ActivityRequestCode requestCode: Int)

        fun showAboutDialog()

        fun showLikeUsDialog()

        fun showRateUsDialog()

        fun showFeedbackDialog()

        fun showZoomWindowSettingsDialog(sharedPreferences: UserPreferences)

        fun showAdvancedSettingsDialog()

        fun showOverwriteDialog(permissionCode: Int, isExport: Boolean)

        fun showPngInformationDialog()

        fun showJpgInformationDialog()

        fun showOraInformationDialog()

        fun showCatrobatInformationDialog()

        fun sendFeedback()

        fun startWelcomeActivity(@ActivityRequestCode requestCode: Int)

        fun startShareImageActivity(bitmap: Bitmap?)

        fun showIndeterminateProgressDialog()

        fun dismissIndeterminateProgressDialog()

        fun returnToPocketCode(path: String?)

        fun showToast(msg: String, duration: Int)

        fun showToast(@StringRes resId: Int, duration: Int)

        fun showSaveErrorDialog()

        fun showLoadErrorDialog()

        fun showRequestPermissionRationaleDialog(
            permissionType: PermissionType,
            permissions: Array<String>,
            requestCode: Int
        )

        fun showRequestPermanentlyDeniedPermissionRationaleDialog()

        fun askForPermission(permissions: Array<String>, requestCode: Int)

        fun doIHavePermission(permission: String): Boolean

        fun isPermissionPermanentlyDenied(permissions: Array<String>): Boolean

        fun finishActivity()

        fun showSaveBeforeFinishDialog()

        fun showSaveBeforeNewImageDialog()

        fun showSaveBeforeLoadImageDialog()

        fun showSaveImageInformationDialogWhenStandalone(
            permissionCode: Int,
            imageNumber: Int,
            isExport: Boolean
        )

        fun restoreFragmentListeners()

        fun showToolChangeToast(offset: Int, idRes: Int)

        fun broadcastAddPictureToGallery(uri: Uri)

        fun rateUsClicked()

        fun showImageImportDialog()

        fun setAntialiasingOnToolPaint()

        fun showCatroidMediaGallery()

        fun showScaleImageRequestDialog(uri: Uri?, requestCode: Int)

        fun setMaskFilterToNull()
    }

    interface MainView {
        val finishing: Boolean
        val isKeyboardShown: Boolean
        val myContentResolver: ContentResolver
        val displayMetrics: DisplayMetrics
        val presenter: Presenter

        fun initializeActionBar(isOpenedFromCatroid: Boolean)

        fun superHandleActivityResult(requestCode: Int, resultCode: Int, data: Intent?)

        fun superHandleRequestPermissionsResult(
            requestCode: Int,
            permissions: Array<String>,
            grantResults: IntArray
        )

        fun getUriFromFile(file: File): Uri

        fun hideKeyboard()

        fun refreshDrawingSurface()

        fun enterHideButtons()

        fun exitHideButtons()

        fun showContentLoadingProgressBar()

        fun hideContentLoadingProgressBar()
    }

    interface Presenter {
        val imageNumber: Int
        val bitmap: Bitmap?
        val context: Context

        fun initializeFromCleanState(extraPicturePath: String?, extraPictureName: String?)

        fun restoreState(
            isFullscreen: Boolean,
            isSaved: Boolean,
            isOpenedFromCatroid: Boolean,
            isOpenedFromFormulaEditorInCatroid: Boolean,
            savedPictureUri: Uri?,
            cameraImageUri: Uri?
        )

        fun finishInitialize()

        fun removeMoreOptionsItems(menu: Menu?)

        fun replaceImageClicked()

        fun addImageToCurrentLayerClicked()

        fun loadNewImage()

        fun newImageClicked()

        fun discardImageClicked()

        fun saveCopyClicked(isExport: Boolean)

        fun saveImageClicked()

        fun saveProjectClicked()

        fun shareImageClicked()

        fun enterHideButtonsClicked()

        fun exitHideButtonsClicked()

        fun backToPocketCodeClicked()

        fun showHelpClicked()

        fun showAboutClicked()

        fun showZoomWindowSettingsClicked(sharedPreferences: UserPreferences)

        fun showAdvancedSettingsClicked()

        fun showRateUsDialog()

        fun showFeedbackDialog()

        fun showOverwriteDialog(permissionCode: Int, isExport: Boolean)

        fun showPngInformationDialog()

        fun showJpgInformationDialog()

        fun showOraInformationDialog()

        fun showCatrobatInformationDialog()

        fun sendFeedback()

        fun onNewImage()

        fun switchBetweenVersions(requestCode: Int, isExport: Boolean)

        fun handleActivityResult(requestCode: Int, resultCode: Int, data: Intent?)

        fun handleRequestPermissionsResult(
            requestCode: Int,
            permissions: Array<String>,
            grantResults: IntArray
        )

        fun onBackPressed()

        fun saveImageConfirmClicked(requestCode: Int, uri: Uri?)

        fun saveProjectConfirmClicked(requestCode: Int, uri: Uri?, imagePreviewUri: Uri?)

        fun saveCopyConfirmClicked(requestCode: Int, uri: Uri?)

        fun undoClicked()

        fun redoClicked()

        fun showColorPickerClicked()

        fun showLayerMenuClicked()

        fun onCommandPostExecute()

        fun setBottomNavigationColor(color: Int)

        fun onCreateTool()

        fun toolClicked(toolType: ToolType)

        fun saveBeforeLoadImage()

        fun saveBeforeNewImage()

        fun saveBeforeFinish()

        fun finishActivity()

        fun actionToolsClicked()

        fun actionCurrentToolClicked()

        fun rateUsClicked()

        fun importFromGalleryClicked()

        fun showImportDialog()

        fun importStickersClicked()

        fun bitmapLoadedFromSource(loadedImage: Bitmap)

        fun setLayerAdapter(layerAdapter: LayerAdapter)

        fun loadScaledImage(uri: Uri?, @ActivityRequestCode requestCode: Int)

        fun setAntialiasingOnOkClicked()

        fun saveNewTemporaryImage()

        fun openTemporaryFile(): WorkspaceReturnValue?

        fun checkForTemporaryFile(): Boolean

        fun setColorHistoryAfterLoadImage(colorHistory: ColorHistory?)
    }

    interface Model {
        var cameraImageUri: Uri?
        var savedPictureUri: Uri?
        var isSaved: Boolean
        var isFullscreen: Boolean
        var isOpenedFromCatroid: Boolean
        var isOpenedFromFormulaEditorInCatroid: Boolean
        var colorHistory: ColorHistory

        fun wasInitialAnimationPlayed(): Boolean

        fun setInitialAnimationPlayed(wasInitialAnimationPlayed: Boolean)
    }

    interface Interactor {
        fun saveCopy(
            callback: SaveImageCallback,
            requestCode: Int,
            layerModel: LayerContracts.Model,
            commandSerializer: CommandSerializer,
            uri: Uri?,
            context: Context
        )

        fun createFile(callback: CreateFileCallback, requestCode: Int, filename: String)

        fun saveImage(
            callback: SaveImageCallback,
            requestCode: Int,
            layerModel: LayerContracts.Model,
            commandSerializer: CommandSerializer,
            uri: Uri?,
            context: Context
        )

        fun saveProject(
            callback: SaveImageCallback,
            requestCode: Int,
            layerModel: LayerContracts.Model,
            commandSerializer: CommandSerializer,
            uri: Uri?,
            imagePreviewUri: Uri?,
            context: Context
        )

        fun loadFile(
            callback: LoadImageCallback,
            requestCode: Int,
            uri: Uri?,
            context: Context,
            scaling: Boolean,
            commandSerializer: CommandSerializer,
        )
    }

    interface TopBarViewHolder {
        val height: Int

        fun enableUndoButton()

        fun disableUndoButton()

        fun enableRedoButton()

        fun disableRedoButton()

        fun hide()

        fun show()

        fun removeStandaloneMenuItems(menu: Menu?)

        fun removeCatroidMenuItems(menu: Menu?)

        fun hideTitleIfNotStandalone()
    }

    interface DrawerLayoutViewHolder {
        fun closeDrawer(gravity: Int, animate: Boolean)

        fun isDrawerOpen(gravity: Int): Boolean

        fun isDrawerVisible(gravity: Int): Boolean

        fun openDrawer(gravity: Int)
    }

    interface BottomBarViewHolder {
        val isVisible: Boolean

        fun show()

        fun hide()
    }

    interface BottomNavigationViewHolder {
        fun show()

        fun hide()

        fun showCurrentTool(toolType: ToolType?)

        fun setColorButtonColor(@ColorInt color: Int)
    }

    interface BottomNavigationAppearance {
        fun showCurrentTool(toolType: ToolType)
    }
}
